### PR TITLE
Firebase cards

### PR DESCRIPTION
**Pull request by annierusu**
Basic functionality of the database with ui

### CODE REVIEW

**Review by Bjolidon — COMMENTED** [review]
Nice code overall with clear comments, just add new lines at the end of the files

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.ktx.database
+import com.google.firebase.ktx.Firebase
+import java.util.concurrent.CompletableFuture
+
+/**
+ * This class is used to manage the user's collection of cards.
+ */
+
+class UserCardsRTDB : AppCompatActivity(){
```
Comment (JulienDiTria) [inline app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:19]: does it need to be an activity? there is no UI associated ?

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    private val db = Firebase.database.reference
+    private val user = Firebase.auth.currentUser
+    private val userCardCollection = db.child(user!!.uid)
+
+    /**
+     * Adds a card to the user's collection.
+     */
+    fun addCardToCollection(card: MagicCard) : String {
+        val cardUID = card.set.code + card.number
+        userCardCollection.child(cardUID).child("name").setValue(card.name)
```
Comment (JulienDiTria) [inline app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:29]: can there be errors while setting those values ?

File: app/src/main/java/com/github/bjolidon/bootcamp/ui/authentication/AuthenticationActivity.kt [hunk 3]
```diff
@@ -11,6 +11,7 @@ import com.firebase.ui.auth.AuthUI.IdpConfig
 import com.firebase.ui.auth.FirebaseAuthUIActivityResultContract
 import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult
 import com.github.bjolidon.bootcamp.R
+import com.github.bjolidon.bootcamp.database.UserCardsActivity
```
Comment (JulienDiTria) [inline app/src/main/java/com/github/bjolidon/bootcamp/ui/authentication/AuthenticationActivity.kt:14]: remove if not used in this file

File: app/src/main/res/layout/activity_sign_out.xml [hunk 4]
```diff
@@ -41,4 +41,15 @@
         app:layout_constraintTop_toTopOf="parent"
         app:layout_constraintVertical_bias="0.802" />
 
+    <Button
+        android:id="@+id/manage_collection_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="52dp"
+        android:text="Manage Collection"
```
Comment (JulienDiTria) [inline app/src/main/res/layout/activity_sign_out.xml:49]: as CI says, put a string resource here

File: app/src/main/res/layout/activity_user_cards.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_bias="0.925" />
+
+    <Button
+        android:id="@+id/set_card_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="448dp"
+        android:text="Add Card"
```
Comment (JulienDiTria) [inline app/src/main/res/layout/activity_user_cards.xml:26]: same here

File: app/src/main/res/layout/activity_user_cards.xml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintHorizontal_bias="0.5"
+        app:layout_constraintStart_toStartOf="parent" />
+
+    <Button
+        android:id="@+id/get_card_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="300dp"
+        android:text="Get"
```
Comment (JulienDiTria) [inline app/src/main/res/layout/activity_user_cards.xml:37]: same

File: app/src/main/res/layout/activity_user_cards.xml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintHorizontal_bias="0.5"
+        app:layout_constraintStart_toStartOf="parent" />
+
+    <TextView
+        android:id="@+id/text_get_card"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="242dp"
+        android:text="get"
```
Comment (JulienDiTria) [inline app/src/main/res/layout/activity_user_cards.xml:48]: same

File: app/src/main/res/layout/activity_user_cards.xml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintHorizontal_bias="0.5"
+        app:layout_constraintStart_toStartOf="parent" />
+
+    <TextView
+        android:id="@+id/text_set_card"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="396dp"
+        android:text="set"
```
Comment (JulienDiTria) [inline app/src/main/res/layout/activity_user_cards.xml:59]: same

**Review by JulienDiTria — CHANGES_REQUESTED** [review]
good work,
curious about the UserCardsRTDB, is it useful or just some work in progress ?

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt (hunk 1 above)
Comment (annierusu) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:19]: oh yeah most likely no, the point of UserCardsRTDB was to hold the functions responsible for fetching the actual data (so nothing to do with ui). This allows these functions to be useful later on if we decide we need to access the database somewhere

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt (hunk 2 above)
Comment (annierusu) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:29]: The code at the moment assumes the card passed to the function is correct. setValue either adds the value and creates a new child if it doesnt exist already under that key, or it overwrites the old value if the key already exists (thats why its important to have unique IDs for each card)

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt (hunk 1 above)
Comment (JulienDiTria) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:19]: then i would make it a class only, not an Activity

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt (hunk 2 above)
Comment (JulienDiTria) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:29]: i understand, but setting this value can not fail ? this is nice !

Comment (JulienDiTria) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:29]: i just checked, the setValue(...) can return a task which can fail or succeed. Will be needed to handle cancellation and failure maybe

Comment (JulienDiTria) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:29]: then put a todo in comment to remember to do it. It s gonna be a task for next sprint

**Review by JulienDiTria — DISMISSED** [review]
tksss

**Review by JulienDiTria — APPROVED** [review]
nice catch

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
```
Comment (AlSchlo) [inline app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:25]: Code redundancies => Refactor this

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsActivity.kt [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        val setMessage: TextView = findViewById(R.id.text_set_card)
+        setMessage.text = ""
+        val getMessage: TextView = findViewById(R.id.text_get_card)
+        getMessage.text = ""
+        var card : MagicCard
+
+        val backButton = findViewById<Button>(R.id.button_back_home2)
+        backButton.setOnClickListener {
+            val intent = Intent(this, MainActivity::class.java)
+            startActivity(intent)
```
Comment (AlSchlo) [inline app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsActivity.kt:56]: Back button should not **restart** a new main activity

Comment (AlSchlo) [reply app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsActivity.kt:56]: You could use the normal Android button for that.

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsActivity.kt [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        setButton.setOnClickListener {
+            val msg = usc.addCardToCollection(card2) //this is for now hardcoded but will be replaced by a function call
+            setMessage.text = msg
+
+        }
+
+        val getButton = findViewById<Button>(R.id.get_card_button)
+
+        getButton.setOnClickListener{
+            //this code should be in a separate function in UserCardsRTDB so it can be separated from this activitiy's UI
```
Comment (AlSchlo) [inline app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsActivity.kt:71]: Indeed it should

File: app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    fun removeCardFromCollection(card: MagicCard) : String {
+        val cardUID = card.set.code + card.number
+        userCardCollection.child(cardUID).removeValue()
+        return "Card $cardUID successfully removed from collection"
+    }
+
+    /**
+     * Transforms the data obtained from the database to a MagicCard object.
+     */
+    fun transformData(data: DataSnapshot): MagicCard{
```
Comment (AlSchlo) [inline app/src/main/java/com/github/bjolidon/bootcamp/database/UserCardsRTDB.kt:51]: Could use a json library. You can send objects directly to your DB (and they are automatically parsed).

**Pull request merged**